Core/Parser/SchemaParser.cs

```diff
@@ -31,6 +32,35 @@ public class SchemaParser
         private readonly string _nameSpace;
         private List<Token> _tokens => _tokenizer.Tokens;
 
+        private void AddDefinition(string name, Definition definition)
+        {
+            _definitions.Add(name, definition);
+            if (_scopes.Count > 0)
+            {
+                _scopes.Last().Add(definition);
+            }
+        }
+
+        /// <summary>
+        /// Open a scope. All definitions added from this point will be added to the scope.
+        /// </summary>
+        private void StartScope() => _scopes.Add(new List<Definition>());
+
+        /// <summary>
+        /// Close the current scope, setting the parent of every enclosed definition to the provided Definition and adding the Definition at the end, in the next scope down.
+        /// </summary>
+        /// <param name="parent"></param>
+        private void CloseScope(string name, Definition parent)
+        {
+            var scope = _scopes.Last();
+            _scopes.Remove(scope);
```

Comment:
This makes me think `_scopes` should be a [`Stack<T>`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.stack-1.pop?view=net-5.0). Maybe that also nicely captures how it is actually used (i.e. even conceptually, it's a stack of scopes).

Reply:
You're right, I'm using that list just like a stack lol. I've got a bad case of javascript-brain where you have to use normal arrays for every kind of linear data structure.